pupil_src/shared_modules/pupil_remote.py

```diff
@@ -67,45 +68,106 @@ class Pupil_Remote(Plugin):
         order (float): See plugin.py
         thread_pipe (zmq.Socket): Pipe for background communication
     """
-    def __init__(self, g_pool,address="tcp://*:50020"):
+    def __init__(self, g_pool,port="50020", host="*", use_primary_interface=True):
         super(Pupil_Remote, self).__init__(g_pool)
         self.order = .01 #excecute first
         self.context = g_pool.zmq_ctx
         self.thread_pipe = zhelper.zthread_fork(self.context, self.thread_loop)
-        self.address = address
-        self.start_server(address)
+
+        self.use_primary_interface = use_primary_interface
+        self.host = host
+        self.port= port
+
+        self.start_server(host+':'+port)
         self.menu = None
 
 
     def start_server(self,new_address):
-        self.thread_pipe.send('Bind')
-        self.thread_pipe.send(new_address)
-        response = self.thread_pipe.recv()
-        if response == 'Bind OK':
-            self.address = new_address
+        self.thread_pipe.send_multipart(('Bind',"tcp://"+new_address))
+        response,msg = self.thread_pipe.recv_multipart()
+        if response == 'Bind OK' :
+            host,port = msg.split(':')
+            self.host = host
+            self.port = port
+            return
+
+        #fail logic
+        logger.error(msg)
+
+        #for service we shut down
+        if self.g_pool.app == 'service':
+            audio.say("Error: Port already in use.")
+            self.notify_all({'subject':'service_process.should_stop'})
+            return
+
+        #for capture we try to bind to a arbitrary port on the first external interface
         else:
-            logger.error(response)
-            if self.g_pool.app == 'service':
-                audio.say("Error: Port already in use.")
-                self.notify_all({'subject':'service_process.should_stop'})
+            self.thread_pipe.send_multipart(('Bind',"tcp://*:*"))
```

Comment:
Generally it would be nice to be able to enumerate network interfaces instead of having the user type ip addresses. Initialisation would look somehow like this:

```python
success = False
if interface_available(last_interface):
  try:
    success = bind(last_interface, last_port)
  except:
    success = bind(last_interface, random_port)
if not success:
  try:
    success = bind(primary_interface, last_port)
  except:
    success = bind(primary_interface, random_port)
if not success:
  logger.error()
  return
```